crates/error/src/ptr.rs

```diff
@@ -0,0 +1,155 @@
+//! Newtype wrappers over raw pointers that document ownership. We use these in
+//! various places instead of safe references because our type-punning would
+//! trigger UB otherwise.
+
+use alloc::boxed::Box;
+use core::{marker::PhantomData, ptr::NonNull};
+
+/// A raw, owned pointer.
+///
+/// You are required to call `T`'s `Drop` and deallocate the pointer, this won't
+/// automatically do it for you like `Box`.
```

Comment:
Would it make sense to implement `Drop` on this struct with a panic, to ensure that we always deconstruct it via `into_*` instead?

Reply:
This gets a little funky in various places, for example in `BoxedDynError`, where we are in a `Drop` implementation and so we only have `&mut self` and are forced to make a raw copy of an `OwnedPtr` in order to turn it into a `Box` and run the inner `Drop` and deallocate it, but then there is still an `OwnedPointer` on the `&mut self`. So we would need to essentially turn that into an `Option<OwnedPointer>`, effectively acting as a drop flag and adding a dynamic check to all uses, which is unnecessary overhead. So I'd prefer not to.

Reply:
OK, yeah, no need to do that. Was mostly wondering if there was a reasonable path to add a little more dynamically-asserted safety here but it's not critical and this is all encapsulated locally anyway.